prov/tcp/src/tcpx_ep.c

```diff
@@ -0,0 +1,769 @@
+
+}
+
+
+static ssize_t tcpx_inject(struct fid_ep *ep, const void *buf, size_t len,
+			   fi_addr_t dest_addr)
+{
+	struct fi_msg msg;
+	struct iovec msg_iov;
+
+	memset(&msg, 0, sizeof(msg));
+	msg_iov.iov_base = (void *) buf;
+	msg_iov.iov_len = len;
+	msg.msg_iov = &msg_iov;
+	msg.iov_count = 1;
+	msg.addr = dest_addr;
+
+	return tcpx_sendmsg(ep, &msg, FI_INJECT | TCPX_NO_COMPLETION);
+
+}
+
+static ssize_t tcpx_senddata(struct fid_ep *ep, const void *buf, size_t len, void *desc,
+			     uint64_t data, fi_addr_t dest_addr, void *context)
+{
+	return -FI_ENOSYS;
+}
+
+static ssize_t tcpx_injectdata(struct fid_ep *ep, const void *buf, size_t len,
+			       uint64_t data, fi_addr_t dest_addr)
+{
+	return -FI_ENOSYS;
+}
+
+
+static struct fi_ops_msg tcpx_msg_ops = {
+	.size = sizeof(struct fi_ops_msg),
+	.recv = tcpx_recv,
+	.recvv = tcpx_recvv,
+	.recvmsg = tcpx_recvmsg,
+	.send = tcpx_send,
+	.sendv = tcpx_sendv,
+	.sendmsg = tcpx_sendmsg,
+	.inject = tcpx_inject,
+	.senddata = tcpx_senddata,
+	.injectdata = tcpx_injectdata,
+};
+
+static void tcpx_set_sockopts(int sock)
+{
+	int optval;
+	optval = 1;
+
+	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)))
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"setsockopt reuseaddr failed\n");
+
+	if (setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, &optval, sizeof(optval)))
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"setsockopt nodelay failed\n");
+
+}
+
+static void *tcpx_ep_connect_handler(void *data)
+{
+	struct tcpx_conn_handle *handle;
+	struct fi_eq_cm_entry *eq_entry;
+	struct tcpx_ep *tcpx_ep = (struct tcpx_ep *) data;
+
+	handle = tcpx_ep->handle;
+
+	eq_entry  = calloc(1, sizeof(*eq_entry));
+	if (!eq_entry) {
+	  FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+		  "cannot allocate memory\n");
+	  goto err;
+	}
+
+	eq_entry->fid = &tcpx_ep->util_ep.ep_fid.fid;
+	eq_entry->info = &tcpx_ep->info;
+
+	tcpx_set_sockopts(handle->conn_fd);
+	if (connect(handle->conn_fd,
+		    (struct sockaddr *)handle->serv_addr,
+		    sizeof(*handle->serv_addr))) {
+
+	  FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+		  "connect failed : %s\n", strerror(errno));
+	  goto err1;
+	}
+	fi_fd_nonblock(handle->conn_fd);
+
+	/* report success to eq with FI_CONNECTED */
+	fi_eq_write(&tcpx_ep->util_ep.eq->eq_fid, FI_CONNECTED,
+		    eq_entry, sizeof(*eq_entry), 0);
+
+
+
+	return NULL;
+err1:
+	free(eq_entry);
+err:
+	ofi_close_socket(handle->conn_fd);
+
+	/* report the error to eq */
+	fi_eq_write(&tcpx_ep->util_ep.eq->eq_fid, FI_SHUTDOWN,
+		    eq_entry, sizeof(*eq_entry), 0);
+
+
+	return NULL;
+}
+
+
+static int tcpx_ep_connect(struct fid_ep *ep, const void *addr,
+		    const void *param, size_t paramlen)
+{
+	struct tcpx_ep *tcpx_ep = container_of(ep, struct tcpx_ep, util_ep.ep_fid);
+	struct sockaddr_in *serv_addr = (struct sockaddr_in *) addr;
+	struct tcpx_conn_handle *handle = tcpx_ep->handle;
+	int ret = 0;
+
+	if (!addr)
+		return -FI_EINVAL;
+
+	handle->conn_fd = socket(AF_INET, SOCK_STREAM, 0);
```

Comment:
We should be able to allocate the socket during EP creation.  This will allow us to bind the source address prior to the user calling connect.